### PR TITLE
settings.py: add 'proxy_images'

### PR DESCRIPTION
**Pull request by zrose584**

### CODE REVIEW

File: youtube/local_playlist.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ def get_local_playlist_videos(name, offset=0, amount=50):
         try:
             info = json.loads(video_json)
             if info['id'] + ".jpg" in thumbnails:
-                info['thumbnail'] = "/youtube.com/data/playlist_thumbnails/" + name + "/" + info['id'] + ".jpg"
+                info['thumbnail'] = settings.img_prefix + "https://youtube.com/data/playlist_thumbnails/" + name + "/" + info['id'] + ".jpg"
```
Comment (user234683) [inline youtube/local_playlist.py:55]: These should always be prefixed because they're requests for local files.

File: youtube/playlist.py [hunk 2]
```diff
@@ -103,7 +104,7 @@ def get_playlist_page():
         util.prefix_urls(item)
         util.add_extra_html_info(item)
         if 'id' in item:
-            item['thumbnail'] = '/https://i.ytimg.com/vi/' + item['id'] + '/default.jpg'
+            item['thumbnail'] = f'{settings.img_prefix}https://i.ytimg.com/vi/' + item['id'] + '/default.jpg'
```
Comment (user234683) [inline youtube/playlist.py:107]: Should be concatenation for consistency

File: youtube/subscriptions.py [hunk 3]
```diff
@@ -820,6 +820,7 @@ def get_subscriptions_page():
             videos, number_of_videos_in_db = _get_videos(cursor, 60, (page - 1)*60, tag)
             for video in videos:
                 video['thumbnail'] = util.URL_ORIGIN + '/data/subscription_thumbnails/' + video['id'] + '.jpg'
+                if not settings.proxy_images: video['thumbnail'] = video['thumbnail'][1:]
```
Comment (user234683) [inline youtube/subscriptions.py:823]: Same issue. These are local files.
Single-line if statements are discouraged (PEP 8).

File: youtube/templates/base.html [hunk 4]
```diff
@@ -3,7 +3,9 @@
     <head>
         <meta charset="utf-8">
         <title>{{ page_title }}</title>
-        <meta http-equiv="Content-Security-Policy" content="default-src 'self' 'unsafe-inline'; media-src 'self' https://*.googlevideo.com">
+        {% if settings.proxy_images %}
+            <meta http-equiv="Content-Security-Policy" content="default-src 'self' 'unsafe-inline'; media-src 'self' https://*.googlevideo.com">
```
Comment (user234683) [inline youtube/templates/base.html:7]: Don't remove the CSP completely, it's there to make sure bugs/exploits don't cause the browser to make non-anonymous requests.

File: youtube/util.py [hunk 5]
```diff
@@ -342,8 +342,8 @@ def video_id(url):
 
 # default, sddefault, mqdefault, hqdefault, hq720
 def get_thumbnail_url(video_id):
-    return "/i.ytimg.com/vi/" + video_id + "/mqdefault.jpg"
-    
+    return f"{settings.img_prefix}https://i.ytimg.com/vi/" + video_id + "/mqdefault.jpg"
```
Comment (user234683) [inline youtube/util.py:345]: Ditto

**Review by user234683 — COMMENTED** [review]
May I ask, what's the intent of the setting (i.e., is the extra complexity really worth it)?

I left some feedback on the code for reference, but this sort of feature would be cleaner if it acted in server.py at the routing level. It should still route through the server itself but not through Tor, so that the browser doesn't send any identifying info to the image servers that it doesn't have to.

`proxy_site` could be replaced with an intermediate handler which checks the setting for `lh3.googleusercontent.com`, `yt3.ggpht.com`, and `ytimg.com` in server.py.

**Comment by zrose584** [comment]
The motivation stems from an addon I use, [imagus](https://addons.mozilla.org/en-US/firefox/addon/imagus/).
It has handmade rules for ggpht.com and ytimg.com.
Porting those to youtube_local would risk inconsistency with the original/upstream rules.

> It should still route through the server itself but not through Tor, so that the browser doesn't send any identifying info to the image servers that it doesn't have to.

No, that would defeat the purpose of this patch. Also I think it's unlikely they use (image) data requests for user identification.
And even if, I could use some header-manipulation addon to achieve what youtube_local does (it's just about the HTTP-Headers, right?).

File: youtube/local_playlist.py (hunk 1 above)
Comment (zrose584) [reply youtube/local_playlist.py:55]: What do you mean by "local files"? Are the thumbnails cached on disk?

**Comment by zrose584** [comment]
The addon uses element's `src` attribute, which is unchanged by a 302.
Can this be merged now?

**Pull request merged**